hazelcast/src/main/java/com/hazelcast/sql/impl/client/QueryClientStateRegistry.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.client;
+
+import com.hazelcast.internal.serialization.Data;
+import com.hazelcast.internal.serialization.InternalSerializationService;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.SqlRow;
+import com.hazelcast.sql.impl.QueryId;
+import com.hazelcast.sql.impl.SqlResultImpl;
+import com.hazelcast.sql.impl.SqlRowImpl;
+import com.hazelcast.sql.impl.row.Row;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Registry of active client cursors.
+ */
+public class QueryClientStateRegistry {
+    /** Registered client cursors. */
+    private final ConcurrentHashMap<QueryId, QueryClientState> clientCursors = new ConcurrentHashMap<>();
```

Comment:
I could not find the cleanup code for the case where a client lefts the cluster without closing the query.
You can use `ClientEndpoint.addDestroyAction(UUID registrationId, Callable<Boolean> removeAction);` for this. 

A client enndpoint is removed and all desrtoy actions are called when the assocaiated connection is disconnected. Since the queries are desinged to be `sticky` it looks like a good fit.

Reply:
And can we have a test to verify that cleanups are done correctly?

Reply:
I see that QueryClientStateRegistry.update is used for this purpose. I think, our standard solution with ClientEndpoint addDestroyAction is enough for this. What do you think ?

Reply:
Please see:
com.hazelcast.sql.impl.client.QueryClientStateRegistry#update
com.hazelcast.sql.SqlErrorClientTest#testCursorCleanupOnClientLeave

Reply:
`ClientEndpoint addDestroyAction` looks heavier to me - one have to register it for every request, and also be careful to deregister it properly. 
The key observation is that majority of requests will be completed normally, without client disconnects. I think the current approach handles it better - we do no do any additional actions for the most queries.

Reply:
Ok. Fair enough.